sqldelight-compiler/src/test/kotlin/com.squareup.sqldelight.types/ResolverTests.kt

```diff
@@ -0,0 +1,98 @@
+import org.antlr.v4.runtime.ParserRuleContext
+import org.junit.Test
+import java.io.File
+import java.io.FileInputStream
+
+class ResolverTests {
+  private val parsed = parse(File("src/test/data/ResolverTestData.sq"))
+  private val symbolTable = SymbolTable(parsed)
+  private val resolver = Resolver(symbolTable)
+
+  @Test
+  fun selectAll() {
+    val resolution = resolver.resolve(parsed.statementWithName("select_all"))
+    assertThat(resolution)
+        .isSelected("_id", "test_column1", "test_column2", tableName = "test")
+  }
+
+  @Test
+  fun selectCount() {
+    val resolution = resolver.resolve(parsed.statementWithName("select_count"))
+    assertThat(resolution)
+        .isSelected("count")
+  }
+
+  @Test
+  fun selectFromSubquery() {
+    val resolution = resolver.resolve(parsed.statementWithName("select_from_subquery"))
+    assertThat(resolution)
+        .isSelected("_id", "test_column1", "test_column2", tableName = "test")
+  }
+
+  @Test
+  fun selectCountFromSubquery() {
+    val resolution = resolver.resolve(parsed.statementWithName("select_count_from_subquery"))
+    assertThat(resolution)
+        .isSelected("count")
+  }
+
+  @Test
+  fun subqueryInWhere() {
+    val resolution = resolver.resolve(parsed.statementWithName("subquery_in_where"))
+    assertThat(resolution)
+        .isSelected("_id", "test_column1", "test_column2", tableName = "test")
+  }
+
+  @Test
+  fun selectFromValues() {
+    val resolution = resolver.resolve(parsed.statementWithName("select_from_values"))
+    assertThat(resolution).hasSize(3)
+    assertThat(resolution).hasSize(3)
+  }
+
+  fun parse(file: File): SqliteParser.ParseContext {
+    FileInputStream(file).use { inputStream ->
+      val lexer = SqliteLexer(ANTLRInputStream(inputStream))
+      lexer.removeErrorListeners()
+
+      val parser = SqliteParser(CommonTokenStream(lexer))
+      parser.removeErrorListeners()
+      return parser.parse()
+    }
+  }
+
+  private fun assertThat(values: List<Value>) = ValuesSubject(values)
+
+  private fun SqliteParser.ParseContext.statementWithName(name: String): ParserRuleContext {
+    val child = sql_stmt_list().sql_stmt().find({ it.sql_stmt_name().text == name })
+    return child?.getChild(child.getChildCount() - 1) as ParserRuleContext
+  }
+
+  private class ValuesSubject(
+      val values: List<Value>
+  ) : IterableSubject<ValuesSubject, Value, List<Value>>(Truth.THROW_ASSERTION_ERROR, values) {
+    fun isSelected(vararg columnNames: String, tableName: String? = null): ValuesSubject {
+      columnNames.forEach { hasSelected(it, tableName) }
+      assertThat(columnNames.size).isEqualTo(values.size)
+      return this
+    }
+
+    fun hasSelected(columnName: String, tableName: String? = null): ValuesSubject {
+      val selected = values.filter {
+        it.columnName != null && it.columnName == columnName
+            && (tableName == null || it.tableName == tableName)
+      }
+      assertThat(selected).isNotEmpty()
```

Comment:
You can put `.named("column $columnName in table $tableName")` here for better error messages when the assert fails.